File: java/smartlist/src/main/java/com/smartlist/module/assessmentblock/AssessmentBlockRepository.java
package com.smartlist.module.assessmentblock;

import com.smartlist.model.Assessment;
import com.smartlist.model.AssessmentBlock;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.CrudRepository;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;

@Repository
public interface AssessmentBlockRepository extends CrudRepository<AssessmentBlock, Integer> {

    @Query("SELECT ab FROM AssessmentBlock ab LEFT JOIN FETCH ab.questions q WHERE ab.id = :assessmentBlockId")
    Optional<AssessmentBlock> findByIdWithQuestions(final int assessmentBlockId);

    @Query("SELECT MAX(ab.sortOrder) FROM AssessmentBlock ab WHERE ab.assessment.id = :assessmentId")
    Integer getMaxSortOrder(final int assessmentId);

    List<AssessmentBlock> findAllByAssessment(final Assessment assessment);

    @Query("SELECT ab, q FROM AssessmentBlock ab LEFT JOIN FETCH ab.questions q WHERE ab.assessment = :assessment")
    List<AssessmentBlock> findAllByAssessmentWithQuestions(final Assessment assessment);

    @Modifying
    @Query("UPDATE AssessmentBlock ab SET ab.sortOrder = ab.sortOrder -1 WHERE ab.sortOrder > :sortOrder AND ab.assessment.id = :assessmentId")
    void updateSortOrder(final int sortOrder, final int assessmentId);

    int countByAssessment(final Assessment assessment);
}

File: javascript/smartlist/smartlist-api-mysql-project/src/modules/assessment-block/assessment-block.repository.ts
import { AssessmentBlock } from 'src/entities/assessment-block.entity';
import { Assessment } from 'src/entities/assessment.entity';
import { EntityRepository, Repository } from 'typeorm';

@EntityRepository(AssessmentBlock)
export class AssessmentBlockRepository extends Repository<AssessmentBlock> {
  async findById(id: number): Promise<AssessmentBlock> {
    return this.findOne(id);
  }

  async findAllByAssessment(
    assessment: Assessment,
  ): Promise<AssessmentBlock[]> {
    return this.find({
      where: { assessment: assessment },
      order: {
        sortOrder: 'ASC',
      },
    });
  }

  async findAllByAssessmentWithQuestions(
    assessment: Assessment,
  ): Promise<AssessmentBlock[]> {
    return this.find({
      where: { assessment: assessment },
      order: {
        sortOrder: 'ASC',
      },
      relations: ['questions'],
    });
  }

  async findByIdWithAssessment(id: number): Promise<AssessmentBlock> {
    return this.findOne(id, {
      relations: ['assessment'],
    });
  }

  async findByIdWithQuestions(id: number): Promise<AssessmentBlock> {
    return this.findOne(id, {
      relations: ['questions'],
    });
  }

  async updateSortOrder(sortOrder: number, assessmentId: number) {
    this.manager.query(
      `UPDATE assessment_block SET sort_order = sort_order - 1 WHERE sort_order > ${sortOrder} and assessment_id=${assessmentId}`,
    );
  }

  async getAssessmentBlockCount(assessmentId: number): Promise<number> {
    const rawData: any = await this.manager.query(
      `SELECT count(id) as blockCount FROM assessment_block WHERE assessment_id=${assessmentId}`,
    );
    return Number(rawData[0].blockCount);
  }

  async getMaxSortOrder(assessmentId: number): Promise<number> {
    const sortOrder: any = await this.manager.query(
      `SELECT max(sort_order) as maxSortOrder FROM assessment_block WHERE assessment_id=${assessmentId}`,
    );
    return sortOrder[0].maxSortOrder;
  }
}
